absorb THD into main cmake build

anderspapitto opened the pull request:
We want to move _C into the same cmake invocation that builds
libcaffe2 and libtorch. However, _C depends on THD and c10d, which in
turn depend on libcaffe2. That means that we can't move _C into that
cmake file unless we do these two first. This change does so.

apaszke reviewed (commented):
Won't this make the build much slower?

anderspapitto commented:
@apaszke why so?

apaszke commented:
Because so far those were built as part of the `build_deps` step which you could skip almost every time you tried to rebuild PyTorch while developing.

anderspapitto commented:
@apaszke that's unaffected. `THD` and `c10d` are still part of `build_deps` after this change.

anderspapitto commented:
I'm still tracking down one bug (in (only) some of our linux CI environments, linking c10d/example fails to find -lgloo and -lgloo_cuda). In the meantime this diff is more-or-less ready for review

soumith commented on setup.py:
```diff
@@ -331,7 +331,7 @@ def run(self):
 dep_libs = [
     'nccl', 'caffe2',
-    'libshm', 'libshm_windows', 'gloo', 'THD', 'c10d',
+    'libshm', 'libshm_windows', 'gloo',
```
afaics, gloo and nccl have to come in at the same time as well.
Gloo and NCCL are dependencies of THD and c10d.

You're going have a harder time dealing with side-effects, than to simply move both gloo and nccl in this step.

soumith commented on tools/build_pytorch_libs.sh:
```diff
@@ -177,7 +177,6 @@ function build() {
 		       -DCMAKE_INSTALL_LIBDIR="$INSTALL_DIR/lib" \
 		       -DCUDA_NVCC_FLAGS="$CUDA_NVCC_FLAGS" \
 		       -DCUDA_DEVICE_DEBUG=$CUDA_DEVICE_DEBUG \
-		       -DCMAKE_PREFIX_PATH="$INSTALL_DIR" \
```
hooray!!!!
But this will break on machine with no system NCCL, because now setup.py builds NCCL, and cmake wont be able to find it.
If in case CMake is able to find it, something quite off is going on, i.e. NCCL (and gloo) is being built by both root cmake and setup.py invoking `build_pytorch_libs.sh` separately to build those libs.

soumith commented on torch/lib/THD/CMakeLists.txt:
```diff
@@ -45,7 +45,14 @@ FIND_PACKAGE(MPI)
 
-FIND_PACKAGE(Caffe2 REQUIRED)
+# I don't know if this branch actually still matters, so I am leaving
+# it in. In particular, there is a neighboring file called build.sh
```
It was used for development. But considering THD is deprecated and soon to disappear, I doubt it's needed.
I'd lean towards removing it. cc: @teng-li to double-check

soumith commented on torch/lib/THD/CMakeLists.txt:
```diff
@@ -140,6 +147,16 @@ ENDIF()
 ADD_LIBRARY(THD STATIC ${all_cpp})
+
+if (CAFFE2_CMAKE_BUILDING_WITH_MAIN_REPO)
+  ADD_DEPENDENCIES(THD caffe2_library)
```
if you ADD_DEPENDENCIES on `caffe2` (instead of `caffe2_library`), it should include the caffe2 directories (which include the dirs below) automatically.
That's the right thing to do, because Caffe2's header paths might keep changing, and you're effectively manually maintaining the paths below

soumith commented on torch/lib/c10d/CMakeLists.txt:
```diff
@@ -127,12 +119,14 @@ target_compile_options(c10d PUBLIC
-if(C10D_USE_CUDA)
-  target_include_directories(c10d PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/../tmp_install/include/THC)
-endif()
+add_dependencies(c10d caffe2_library)
```
same comments as for THD, depend on `caffe2`, and remove the manual include_directories section below

soumith reviewed (dismissed):
see inline comments.

Main things to be fixed:

- dont manually list out all the include_directories, directly depend on `caffe2` (and possibly `caffe2_gpu`?
- you very likely have to move `nccl` and `gloo` building into cmake in this PR as well, or have to debug some nasty side-effects

anderspapitto commented:
- I've deferred removal of CMAKE_PREFIX_PATH out of this PR, in order to keep the scope minimal
- I've changed the dependencies on caffe2_library to be dependencies on caffe2
- I've removed a couple include directories. The remainder are necessary, and I've annotated each one with what breaks if you take it out.
- Adding a dependency on caffe2_gpu does nothing
- It's not necessary to involve gloo and/or nccl in this PR. They are built and provided to THD/C10D in the exact same manner as before.

I'm eager to revisit some of these things, because they can definitely be improved. However there are many many tangentially-related improvements to get bogged down in, so I'm tightly limiting the scope of each change.

@soumith

anderspapitto commented on cmake/Dependencies.cmake:
```diff
@@ -442,7 +442,7 @@ if(USE_MPI)
     include_directories(SYSTEM ${MPI_CXX_INCLUDE_PATH})
     list(APPEND Caffe2_DEPENDENCY_LIBS ${MPI_CXX_LIBRARIES})
-    set(CMAKE_EXE_LINKER_FLAGS ${MPI_CXX_LINK_FLAGS})
+    set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} ${MPI_CXX_LINK_FLAGS}")
```
this line is https://github.com/pytorch/pytorch/pull/12834

anderspapitto commented:
The CI failures are all unrelated caffe2 tests. This PR is 100% functional and ready to ship

ezyang commented on torch/lib/THD/CMakeLists.txt:
```diff
@@ -140,6 +139,15 @@ ENDIF()
+  ${CMAKE_BINARY_DIR}/aten/src # provides "ATen/TypeExtendedInterface.h" to ATen.h
+  ${CMAKE_BINARY_DIR}/caffe2/aten/src # provides <TH/THGeneral.h> to THC.h
+  ${CMAKE_BINARY_DIR}/caffe2/aten/src/THC # provides "THCGeneral.h" to THC.h
+  )
```
Mumble mumble modern style cmake should set these include dirs

soumith replied:
what Ed is referring to, and what I mentioned with my comments before is that it's the responsibility of `caffe2` cmake target to automatically include these directories when THD is being compiled, because you declared that THD depends on `caffe2` target via `ADD_DEPENDENCIES(THD caffe2)`

Yangqing replied:
Yes, exactly like Ed and Soumith mentioned. Also, looking at the folders, it seems that when we set the modern inclusion paths, we should use build interface and install interface separately, see this link for details:

https://cmake.org/cmake/help/v3.0/command/target_include_directories.html

anderspapitto replied:
btw I've pulled this cleanup out as a separate task, since it's not a real blocker https://github.com/pytorch/pytorch/issues/12945

ezyang commented:
@pjh5 Can you check if nightlies still work with this?

ezyang reviewed (dismissed):
I think the test failures are real. I did not see these failures on master, and they look 100% reproduceable.

anderspapitto committed:
absorb THD into the main cmake build

soumith commented on torch/lib/THD/CMakeLists.txt:
```diff
@@ -155,10 +160,8 @@ IF(MPI_FOUND)
-  INCLUDE_DIRECTORIES(${GLOO_INCLUDE_DIR})
-  FILE(APPEND "${CMAKE_INSTALL_PREFIX}/THD_deps.txt" "${Gloo_LIBRARY};")
-  FILE(APPEND "${CMAKE_INSTALL_PREFIX}/THD_deps.txt" "${Gloo_NATIVE_LIBRARY};")
+IF(USE_GLOO AND USE_CUDA)
```
GLOO is used for CPU-only backend as well right?
isn't it `IF(USE_GLOO) ADD_DEPENDENCIES(THD gloo)`

soumith commented on torch/lib/THD/CMakeLists.txt:
```diff
@@ -115,7 +108,7 @@ IF(NOT MPI_FOUND)
 ENDIF()
 
-IF(NOT GLOO_FOUND)
+IF(NOT USE_GLOO OR NOT USE_CUDA)
```
see comment below. I think gloo is used for CPU-only parts as well

soumith commented on torch/lib/THD/CMakeLists.txt:
```diff
@@ -67,9 +63,8 @@ IF(MPI_FOUND)
 ENDIF()
 
-IF(GLOO_FOUND)
+IF(USE_GLOO AND USE_CUDA)
```
see comment below. I think gloo is used for CPU-only parts as well

soumith replied:
after reviewing a bit deeper, this indeed has to just be IF(USE_GLOO)

soumith commented on torch/lib/THD/CMakeLists.txt:
```diff
@@ -1,7 +1,6 @@
 CMAKE_MINIMUM_REQUIRED(VERSION 2.8)
 # TODO(jiayq): once we have unified CMake entry, remove this module path.
 SET(CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR}/../../../cmake/Modules ${CMAKE_MODULE_PATH})
-SET(Gloo_USE_CUDA TRUE)
```
this was a harmless line that didn't actually do anything. I checked downstream consequences of this, and there were actually none. Nice that you cleaned up / removed.

soumith replied:
as discussed IRL this wasn't harmless, this basically made sure FIND_PACKAGE(Gloo) returned true only when Gloo was compiled with CUDA.

soumith reviewed (approved):
approved based on further investigations. Gloo is only used in THD when it's built with CUDA support.
